### DISCUSSION TITLE
make apparent startup (almost) instantaneous by keeping running instances under tmux?

### DISCUSSION THREAD

**Discussion by bkerin** [akinomyoga/ble.sh#283] [Ideas]
I'm really liking ble.sh so far but there is one issue that's still interfering with regular use for me and that's the startup time.  It's only about 5 seconds but I use dwm and have habit of often creating new windows with a terminal when investigating side issues and the like, at 5 seconds it becomes annoying.

I thought of using tmux to just start a pool of instances and then attach to them from a bare-bones shell, it seems to work pretty well like so:

```
tmux new-session -d -sshell_pool_1  # ultimately maybe from .xsession or somewhere later
xfce4-terminal --execute ~/local/bin/tmux attach-session -t shell_pool_1     # would ultimately need to get session name from pool to use somehow
```

I think a pool server should be a quick hack (tm) to do but thought I would ask if something like it has already been done or there is some better way already?

Also there is one small visual artifact that suggest some sort of race or something is going on (see attached screenshot).  A copy of the status bar appears in the middle of the terminal for some reason.

![screenshot](https://user-images.githubusercontent.com/7434164/220214061-6534fad3-84f0-499c-a1d7-7dd074ff70fe.jpg)

Even a tiny delay hides this issue:

`xfce4-terminal --execute bash -c "sleep 0.05; ~/local/bin/tmux attach-session -t shell_pool_1"
`

Also the terminal is a tiny bit slower to vanish again than I would like, I think because the [ble: exit] indicator.  I would swear I saw a way to remove that, but the only two options I found going back through is exec_errexit_mark= which only affects behavior of commands, not the mark for exit of ble.sh itself.  builtin exit isn't quite enough either as there is some trap running.  kill -9 $$ works but seems cruel :).  I didn't find out what in the ble.sh exit is pausing things.

**Comment by akinomyoga** [comment]
> I would ask if something like it has already been done or there is some better way already?

It might be interesting to think about pooling the instances, but I'm not sure if that is the correct way to solve the present problem because the 5-second delay seems unusual in my experience. For example, the startup is about 0.2 seconds with my settings of `ble.sh` and other stuff in bashrc. The *better way* depends on the current situation.

- **Q1**: Do you have an idea why there is such a long startup delay? Is it caused by the loading of `ble.sh` itself? Or is it caused by the settings of `ble.sh`? Or is it caused by other settings in your bashrc? I assume that you are using Bash 5.0 as reported in [#278](https://github.com/akinomyoga/ble.sh/issues/278). Could you paste the result of the following commands?

```bash
$ PS1='\$ ' bash --norc # start a child Bash session with a plain setting
$ echo $EPOCHREALTIME; source /path/to/ble.sh --norc --attach=attach; echo $EPOCHREALTIME
$ exit
$ PS1='\$ ' bash --norc # start another Bash session with a plain setting
$ echo $EPOCHREALTIME; source /path/to/ble.sh --attach=attach; echo $EPOCHREALTIME
$ exit
```

If the delay is not caused by the loading of `ble.sh` itself, `ble.sh` has a mechanism to delay some of the slow initialization and make the startup faster (`ble-import -d` or `ble/util/idle.push`), so I can explain it to you. If the delay is caused by the loading of `ble.sh`, I'll tell you how to turn on `ble.sh`'s diagnosis on the startup.

> Also there is one small visual artifact that suggest some sort of race or something is going on (see attached screenshot). A copy of the status bar appears in the middle of the terminal for some reason.

- **Q2**: What is that status bar? Is it the status bar of `ble.sh` set up with `bleopt prompt_status_bar`? Or is it the status bar of `tmux`? How do you configure the status bar?

If it is the status bar set up by `bleopt prompt_status_bar`, I think you can delay the setting, (which is also possible with `ble-import -d` / `ble/util/idle.push`).

> Also the terminal is a tiny bit slower to vanish again than I would like,

In the exit phase, `ble.sh` does non-trivial things including updating the command-history file (such as `~/.bash_history`) and cleaning up temporary files that it uses, but I'm not sure if that causes the large delay. In my environment, it is about 0.17 seconds.

- **Q3**: How long does it take?

> I think because the [ble: exit] indicator.

I don't think the exit delay is caused by the `[ble: exit]` indicator.

> kill -9 $$ works but seems cruel :).

With `kill -9 $$`, there is no chance for Bash/`ble.sh` to save the command history, and also no chance to clean up temporary files and other child processes.

**Reply by akinomyoga** [reply]
@bkerin By the way, in GitHub Discussion, could you reply to the comment thread rather than making a separate thread every time you reply?

----

**edit**: I think the GitHub interface is strange, but you need to reply on the web page rather than replying via email.

https://github.com/community/community/discussions/3287

**Reply by bkerin** [reply]
> > I would ask if something like it has already been done or there is some better way already?
> 
> It might be interesting to think about pooling the instances, but I'm not sure if that is the correct way to solve the present problem because the 5-second delay seems unusual in my experience. For example, the startup is about 0.2 seconds with my settings of `ble.sh` and other stuff in bashrc. The _better way_ depends on the current situation.
> 
> * **Q1**: Do you have an idea why there is such a long startup delay? Is it caused by the loading of `ble.sh` itself? Or is it caused by the settings of `ble.sh`? Or is it caused by other settings in your bashrc? I assume that you are using Bash 5.0 as reported in [#278](https://github.com/akinomyoga/ble.sh/issues/278). Could you paste the result of the following commands?
> 
> ```shell
> $ PS1='\$ ' bash --norc # start a child Bash session with a plain setting
> $ echo $EPOCHREALTIME; source /path/to/ble.sh --norc --attach=attach; echo $EPOCHREALTIME
> $ exit
> $ PS1='\$ ' bash --norc # start another Bash session with a plain setting
> $ echo $EPOCHREALTIME; source /path/to/ble.sh --attach=attach; echo $EPOCHREALTIME
> $ exit
>```

$ echo $EPOCHREALTIME; source ~/.local/share/blesh/ble.sh --norc --attach=attach; echo $EPOCHREALTIME
1677026300.199086
$ 1677026300.494834

$ echo $EPOCHREALTIME; source ~/.local/share/blesh/ble.sh --attach=attach; echo $EPOCHREALTIME
1677026611.866025
$ 1677026612.166080

So it's something in my .bashrc.  It's fast by itself but apparently induces ble.sh to do something slow.  Exit is also very fast with `--norc`.  I'll isolate it hopefull later today or tomorrow.

> If the delay is not caused by the loading of `ble.sh` itself, `ble.sh` has a mechanism to delay some of the slow initialization and make the startup faster (`ble-import -d` or `ble/util/idle.push`), so I can explain it to you. If the delay is caused by the loading of `ble.sh`, I'll tell you how to turn on `ble.sh`'s diagnosis on the startup.
> 
> > Also there is one small visual artifact that suggest some sort of race or something is going on (see attached screenshot). A copy of the status bar appears in the middle of the terminal for some reason.
> 
> * **Q2**: What is that status bar? Is it the status bar of `ble.sh` set up with `bleopt prompt_status_bar`? Or is it the status bar of `tmux`? How do you configure the status bar?

It's the status bar of ble.sh that gets the extra draw.  In the screenshot it's the gray bar with ~.  It's set with `
bleopt prompt_status_line='\w'`.
> 
> If it is the status bar set up by `bleopt prompt_status_bar`, I think you can delay the setting, (which is also possible with `ble-import -d` / `ble/util/idle.push`).
> 
> > Also the terminal is a tiny bit slower to vanish again than I would like,
> 
> In the exit phase, `ble.sh` does non-trivial things including updating the command-history file (such as `~/.bash_history`) and cleaning up temporary files that it uses, but I'm not sure if that causes the large delay. In my environment, it is about 0.17 seconds.
> 
> * **Q3**: How long does it take?
> 
> > I think because the [ble: exit] indicator.
> 
> I don't think the exit delay is caused by the `[ble: exit]` indicator.
> 
> > kill -9 $$ works but seems cruel :).
> 
> With `kill -9 $$`, there is no chance for Bash/`ble.sh` to save the command history, and also no chance to clean up temporary files and other child processes.

It takes about a second in my slow setup.  I'll determine what in my existing .bashrc is causing this also.

**Reply by bkerin** [reply]
Well with pretty much everything removed I started looking for other causes and it looks like this non-branded external battery I just got is getting sensed (or not sensed) and causing some major CPU throttling.  Congrats to ble.sh for being the first thing to work computer hard enough to show the problem :)

**Reply by akinomyoga** [reply]
> So it's something in my .bashrc. It's fast by itself but apparently induces ble.sh to do something slow.

Thank you for the information. If the delay is caused in ble.sh by some interference with the other configurations, I'll take a look if there is sufficient information to reproduce the problem.

> Exit is also very fast with --norc. I'll isolate it hopefull later today or tomorrow.

If something is making ble.sh slow, one possibility would be through `PROMPT_COMMAND` or `PS1`. Another possibility might be the `DEBUG` trap.

> Well with pretty much everything removed I started looking for other causes and it looks like this non-branded external battery I just got is getting sensed (or not sensed) and causing some major CPU throttling. Congrats to ble.sh for being the first thing to work computer hard enough to show the problem :)

Yeah, admittedly, ble.sh consumes much more computational resources compared to plain Bash. I think the CPU usage wouldn't be a real issue, but the large memory footprint is one of the recent problems of `ble.sh` with the latest version of Bash. Somehow, the footprint becomes larger and larger in the newer version of Bash even if we load the same version of `ble.sh`.

**Comment by bkerin** [comment]
Since maybe other users will find ble.sh reveals a throttling problem here is a pointer to what worked for me.  It's probably most likely to solve the issue for Dell laptops but maybe others do the same tricks.

The answer is here *except* you should and with 0xFFFFFFFE (issue reported already):

https://github.com/yyearth/turnoff-BD-PROCHOT

so to see the value to set: 

```
printf '%x\n' $((0x$(rdmsr 0x1FC) & 0xFFFFFFFE))
```

or to script it:

```
wrmsr 0x1fc $(printf '0x%x\n' $((0x$(rdmsr 0x1FC) & 0xFFFFFFFE)))
```

Top guy here is right, it's just a scam from dell to get money from you I've had exactly the issue he describes with a power supply previously and this is just the same garbage for non-dell external batteries (the one I have has greater current capacity the dell external batteries and still it use PROCHOT to hurt them):

https://www.quora.com/What-is-BD-PROCHOT-that-causes-the-CPU-to-throttle-What-can-I-do-to-permanently-disable-it

**Reply by akinomyoga** [reply]
Hmm, that's an interesting story. Thank you for sharing the tip.
